### PR TITLE
feat: support flash attn

### PR DESCRIPTION
**Pull request by ustcfy**
## Motivation

This PR adds flash attn for training, supporting longer ctx and faster training. And the impact on precision is relatively small.
￼￼
image

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by FlamingoPg** [comment]
Cool, any speed/memory compare?

**Comment by ustcfy** [comment]
> Cool, any speed/memory compare?

I used the benchmark from https://github.com/sgl-project/SpecForge/pull/97 for comparison(I removed `attention_mask[:, 128:] = False`). I think my implementation is not good enough. It's only for personal use right now.(
```
==================================================
COMPARISON SUMMARY - Forward + Backward
==================================================
Speed comparison (Eagle vs Flex):
  Eagle: 1.5859s
  Flex:  2.4252s
  Ratio: 0.65x (Eagle is faster)

Memory comparison (Eagle vs Flex):
  Eagle peak: 2.567 GB
  Flex peak:  0.784 GB
  Ratio: 3.27x (Flex uses less)
```
```
==================================================
COMPARISON SUMMARY - Forward + Backward
==================================================
Speed comparison (Eagle vs Flash):
  Eagle: 1.5926s
  Flash: 1.5351s
  Ratio: 1.04x (Flash is faster)

Memory comparison (Eagle vs Flash):
  Eagle peak: 2.567 GB
  Flash peak: 0.846 GB
  Ratio: 3.03x (Flash uses less)
  ```

**Comment by FrankLeeeee** [comment]
Hi, can you add a test file to verify that flash attention does  not lose much precision as part of unit testing practice?

**Comment by yubofredwang** [comment]
This PR looks promising. Currently Flex attention has accuracy issues for bloat16 testing. Can you change the test to bfloat16 and confirm the backward gradients are close?

File: specforge/modeling/draft/llama3_eagle.py [hunk 1]
```diff
@@ -386,6 +415,98 @@ def forward(
...
+        attn_output, lse, _ = flash_attn_func(
+            q,
+            k,
+            v,
+            causal=True,
+            return_attn_probs=True,
+        )
+        lck = len(cache_k)
+
+        attn_weights = lse[..., None]
```
Comment (yubofredwang) [inline specforge/modeling/draft/llama3_eagle.py:480]: I think we are missing the handling of shrinking attention mask here:

`attn_weights = attn_weights + attention_mask`

I think in each iteration the mask shrinks by a diagnol due to https://github.com/sgl-project/SpecForge/blob/e17bdef163d60d7d37af8c32afbeab6b8669aca7/specforge/core/eagle3.py#L413C15-L415C89

Comment (SunnyLi2015) [reply specforge/modeling/draft/llama3_eagle.py:480]: May I know why we need the mask shrink? It doesn't align with inference attention mask, right?
